### DISCUSSION TITLE
BUG: CryptoPkg can't handle certificates over 4KB

### DISCUSSION THREAD

**Discussion by bohmannn** [tianocore/edk2#3646] [General]
Certificates in size over 4KB is not converting
Trying this piece of code:

  UINT8* cert = ... (loaded certificate over 4KB)
  BIO *bio  = NULL;

  bio = BIO_new(BIO_s_mem());

  if (BIO_puts(bio, cert) <= 0)  // This fails
  {

  }

After investigating I've found that it happens because openssl uses function strlen() to read certificate length that mapped to function in file CryptoPkg\Library\Include\CrtLibSupport.h  :

#define MAX_STRING_SIZE  0x1000

#define strlen(str)                       (size_t)(AsciiStrnLenS(str,MAX_STRING_SIZE))

where MAX_STRING_SIZE  is 4096 bytes so openssl can't handle files over this border.

This is weird, there must be unsafe version of AsciiStrLen with unlimited size..

**Comment by Flickdm** [comment]
Out of curiosity, does the certificate you're trying to use have a key length of 4KB (32768 bits) or 4Kb (4096 bits)?

**Reply by bohmannn** [reply]
4096 bytes (32768 bits)

**Reply by Flickdm** [reply]
Sorry for the late reply, I think that's your problem. Typically, 4K keys means _4K bits_ not bytes.
